### PR TITLE
fix 'ready for review'/'published' status bug

### PR DESCRIPTION
**Pull request by chris48s**
This PR is a fix for an issue reported on slack.

Basically the issue here is that we were just looking at

- Is this `LanguageCloudFile` object imported: Yes/No?
- Is the target page it is pointing at published: Yes/No?

This meant that as we translated the same page on multiple occasions over time the status for old `LanguageCloudFile` objects was changing.

This PR attempts to address that by attaching a FK to a `PageRevision` at the point we import the `LanguageCloudFile` so we can accurately track whether the changes related to that specific `LanguageCloudFile` are live or not. I've also added a less specific fallback status to handle some edge cases. The key one being for historic `LanguageCloudFile` objects that don't have a revision attached to them.

side issue: I seem to have hit the same issue Karl did over in https://github.com/wagtail/wagtail-localize/pull/549 with `0066_collection_management_permissions`. I followed the advice from https://github.com/wagtail/wagtail-localize/pull/549#issuecomment-1119600530 and just dropped it but it would be useful to understand why Django thinks that needs to be applied first when really it doesn't.

### CODE REVIEW

**Comment by zerolab** [comment]
Looks like you're using a more recent version of Wagtail which has `0066_collection_management_permissions` (2.16, I believe), so Django happily takes that as the latest.. The dark side of supporting loads of versions..

File: wagtail_localize_rws_languagecloud/models.py [hunk 1]
```diff
@@ -111,11 +117,41 @@ def is_failed(self):
...
+            1. For legacy reasons: we didn't always save a revision on this object.
+               Objects created with <=0.8.0 willl have a NULL revision
+            2. If we threw an error trying to call save_target() in import_po()
+
+            In this case, say we've got the translations back
+            but we don't know if they are published or not
+            """
+            return RECEIVED
+
+        if (
```
Comment (zerolab) [inline wagtail_localize_rws_languagecloud/models.py:143]: nitpick/suggestion: add comments before the next 2 `if`s. imho will help with the cognitive load for future us.

Took me a bit to go over the various scenarios, trying to see whether anything was missed.

So this is: "The imported file is for the latest translation revision, and the translation doesn't have any changes, so show as published"

The other being: "The imported file came from an earlier translation revision, so show as published"

not blocking, and could live as a comment here

File: wagtail_localize_rws_languagecloud/tests/test_models.py [hunk 2]
```diff
@@ -76,7 +76,33 @@ def test_translations_ready(self):
         self.file.save()
         self.assertEqual(self.file.combined_status, "Translations ready for review")
 
-    def test_translations_published(self):
+    def test_translations_no_revision(self):
```
Comment (zerolab) [inline wagtail_localize_rws_languagecloud/tests/test_models.py:79]: suggestion: add a test for the translation.has_unpublished_changes (expecting "Translations ready for review")

File: wagtail_localize_rws_languagecloud/models.py [hunk 3]
```diff
@@ -111,11 +117,41 @@ def is_failed(self):
...
+            self.revision == instance.get_latest_revision()
+            and not instance.has_unpublished_changes
+        ):
+            return PUBLISHED
+
+        if (self.revision != instance.get_latest_revision()) and (
+            self.revision.created_at < instance.get_latest_revision().created_at
```
Comment (zerolab) [inline wagtail_localize_rws_languagecloud/models.py:150]: question: could there ever be a case where the created_at for the revision recorded against the file would not be <= instance.get_latest_revision()?

**Review by zerolab — CHANGES_REQUESTED** [review]

File: wagtail_localize_rws_languagecloud/tests/test_models.py (hunk 2 above)
Comment (chris48s) [reply wagtail_localize_rws_languagecloud/tests/test_models.py:79]: I think this is already covered by the existing test `test_translations_ready`

File: wagtail_localize_rws_languagecloud/models.py (hunk 1 above)
Comment (chris48s) [reply wagtail_localize_rws_languagecloud/models.py:143]: Yes I will add a couple of comments before I merge it (lets see if I need to tweak the logic slightly first)

File: wagtail_localize_rws_languagecloud/models.py (hunk 3 above)
Comment (chris48s) [reply wagtail_localize_rws_languagecloud/models.py:150]: Essentially the condition I am/was trying to express here is "a more recent revision of `instance` than `self.revision` exists".

However, reading this again with fresh eyes it occurs to me there's actually 2 distinct cases to consider here:

- One or more revisions of `instance` which are more recent than `self.revision` exist and one or more of those revisions is published
- One or more revisions of `instance` which are more recent than `self.revision` exist and none of those revisions are published

I think in the first case we label that as "Translations published"
For the second one I wonder if that is another case of "Translations ready for review" or another different (but somewhat difficult to summarise) status?

Comment (zerolab) [reply wagtail_localize_rws_languagecloud/models.py:150]: May I suggest you pose that question to Lucas and the team?

Comment (chris48s) [reply wagtail_localize_rws_languagecloud/models.py:150]: Have posted it on slack. I'll see where that goes then update this.

**Comment by chris48s** [comment]
I've pushed a couple more commits. I think this resolves everything in line with the review comments/slack conversations and also makes the logic a bit easier to understand. Are you able to have another quick look over this @zerolab ? Thanks

**Review by zerolab — APPROVED** [review]
LGTM

**Pull request merged**